From 2864d0266b64420f4bd0cc551920843fa9a5c83b Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Joan=20Bruguera=20Mic=C3=B3?= <joanbrugueram@gmail.com>
Date: Sat, 1 Nov 2025 14:11:25 +0000
Subject: [PATCH] Fix build for Linux Next-20251031

THIS PATCH IS UNTESTED!
TODO: REMOVE pr_warn_ratelimited USES ONCE TESTED.
---
 common/inc/nv-memdbg.h |  5 +++--
 common/inc/nv-time.h   | 10 ++++++++--
 nvidia/nv-dma.c        |  8 ++++++++
 nvidia/os-interface.c  |  2 +-
 4 files changed, 20 insertions(+), 5 deletions(-)

diff --git a/common/inc/nv-memdbg.h b/common/inc/nv-memdbg.h
index c618ead..8212d44 100644
--- a/common/inc/nv-memdbg.h
+++ b/common/inc/nv-memdbg.h
@@ -28,8 +28,9 @@ void nv_memdbg_exit(void);
 
 #else
 
-#define NV_MEMDBG_ADD(ptr, size)
-#define NV_MEMDBG_REMOVE(ptr, size)
+// NB: Using while(0) to avoid -Wempty-body warnings
+#define NV_MEMDBG_ADD(ptr, size) while(0)
+#define NV_MEMDBG_REMOVE(ptr, size) while(0)
 
 #endif /* NV_MEM_LOGGER */
 
diff --git a/common/inc/nv-time.h b/common/inc/nv-time.h
index 0133383..584633f 100644
--- a/common/inc/nv-time.h
+++ b/common/inc/nv-time.h
@@ -31,6 +31,12 @@
 
 #include <nvstatus.h>
 
+#include <linux/version.h>
+#if LINUX_VERSION_CODE < KERNEL_VERSION(5, 11, 0)
+// Rel. commit "treewide: Remove in_irq()" (Matthew Wilcox, 24 Oct 2025)
+#define in_hardirq in_irq
+#endif
+
 #define NV_MAX_ISR_DELAY_US           20000
 #define NV_MAX_ISR_DELAY_MS           (NV_MAX_ISR_DELAY_US / 1000)
 #define NV_NSECS_TO_JIFFIES(nsec)     ((nsec) * HZ / 1000000000)
@@ -141,7 +147,7 @@ static inline NV_STATUS nv_sleep_us(unsigned int us)
     ktime_get_real_ts64(&tm1);
 #endif
 
-    if (in_irq() && (us > NV_MAX_ISR_DELAY_US))
+    if (in_hardirq() && (us > NV_MAX_ISR_DELAY_US))
         return NV_ERR_GENERIC;
 
     mdelay_safe_msec = us / 1000;
@@ -187,7 +193,7 @@ static inline NV_STATUS nv_sleep_ms(unsigned int ms)
     tm_start = tm_aux;
 #endif
 
-    if (in_irq() && (ms > NV_MAX_ISR_DELAY_MS))
+    if (in_hardirq() && (ms > NV_MAX_ISR_DELAY_MS))
     {
         return NV_ERR_GENERIC;
     }
diff --git a/nvidia/nv-dma.c b/nvidia/nv-dma.c
index 9f1ea11..955b044 100644
--- a/nvidia/nv-dma.c
+++ b/nvidia/nv-dma.c
@@ -12,6 +12,7 @@
 
 #include "os-interface.h"
 #include "nv-linux.h"
+#include <linux/version.h>
 
 #define NV_DMA_DEV_PRINTF(debuglevel, dma_dev, format, ... )                \
     nv_printf(debuglevel, "NVRM: %s: " format,                              \
@@ -764,7 +765,14 @@ static NvBool nv_dma_is_map_resource_implemented
 #endif
     }
 
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(999, 0, 0)
+    // Rel. commit "dma-mapping: remove unused mapping resource callbacks" (Leon Romanovsky, 15 Oct 2025)
+    // https://lore.kernel.org/all/20251015-remove-map-page-v5-0-3bbfe3a25cdf@kernel.org/
+    pr_warn_ratelimited("NVIDIA-470xx: Hit ops->map_phys check %s:%d\n", __FILE__, __LINE__);
+    return (ops->map_phys != NULL);
+#else
     return (ops->map_resource != NULL);
+#endif
 #else
     return NV_FALSE;
 #endif
diff --git a/nvidia/os-interface.c b/nvidia/os-interface.c
index e7c1f9d..7379965 100644
--- a/nvidia/os-interface.c
+++ b/nvidia/os-interface.c
@@ -229,7 +229,7 @@ NvBool NV_API_CALL os_semaphore_may_sleep(void)
 
 NvBool NV_API_CALL os_is_isr(void)
 {
-    return (in_irq());
+    return (in_hardirq());
 }
 
 // return TRUE if the caller is the super-user
-- 
2.51.2

